presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java

```diff
@@ -798,62 +787,59 @@ private void transitionToFinished()
     {
         queryStateTimer.endQuery();
 
-        queryState.setIf(FINISHED, currentState -> !currentState.isDone());
+        setQueryStateIf(FINISHED, currentState -> !currentState.isDone());
     }
 
     public boolean transitionToFailed(Throwable throwable)
+    {
+        return transitionToFailed(throwable, false);
+    }
+
+    private boolean transitionToFailed(Throwable throwable, boolean finishingFailed)
     {
         cleanupQueryQuietly();
-        queryStateTimer.endQuery();
 
-        // NOTE: The failure cause must be set before triggering the state change, so
-        // listeners can observe the exception. This is safe because the failure cause
-        // can only be observed if the transition to FAILED is successful.
-        requireNonNull(throwable, "throwable is null");
-        failureCause.compareAndSet(null, toFailure(throwable));
-
-        boolean failed = queryState.setIf(FAILED, currentState -> !currentState.isDone());
-        if (failed) {
-            QUERY_STATE_LOG.debug(throwable, "Query %s failed", queryId);
-            session.getTransactionId().ifPresent(transactionId -> {
-                if (transactionManager.isAutoCommit(transactionId)) {
-                    transactionManager.asyncAbort(transactionId);
-                }
-                else {
-                    transactionManager.fail(transactionId);
-                }
-            });
-        }
-        else {
-            QUERY_STATE_LOG.debug(throwable, "Failure after query %s finished", queryId);
-        }
+        do {
+            QueryStateWrapper observedState = queryState.get();
+
+            if (observedState.isDone()) {
+                QUERY_STATE_LOG.debug(throwable, "Query %s tried to fail while %s", queryId, observedState.unwrap());
```

Comment:
```suggestion
                QUERY_STATE_LOG.debug(throwable, "Failure after query %s finished", queryId);
```

Reply:
I tried to keep every log line beginning the same way (Query ID ...). Does it make sense?